File: lua/hamming/hamming.lua
local hamming = {}

function hamming.compute(first, second)
    assert(#first == #second, "strands must be of equal length")

    local diff = 0
    for i = 1, #first do
        if first:sub(i, i) ~= second:sub(i, i) then
            diff = diff + 1
        end
    end
    return diff
end

return hamming

File: java/hamming/src/main/java/Hamming.java
import java.util.Iterator;

class Hamming {

    private String left;
    private String right;

    Hamming(String leftStrand, String rightStrand) {
        if (leftStrand.length() != rightStrand.length())
            throw new IllegalArgumentException("strands must be of equal length");

        left = leftStrand;
        right = rightStrand;
    }

    int getHammingDistance() {
        // int distance = 0;
        // for (int i = 0; i < left.length(); i++) {
        //     if (left.codePointAt(i) != right.codePointAt(i))
        //         distance++;
        // }
        // return distance;

        Iterator<Integer> rChars = right.codePoints().iterator();
        return (int) left
            .codePoints()
            .filter(c -> c != rChars.next())
            .count();
    }
}
